### PR TITLE
CRenderCapture: split into separate classes

### PR DESCRIPTION
**Pull request by lrusak**
This splits the `CRenderCapture` code into separate classes for each render system (GL/GLES/DX).

This allows the code to be more readable and expandable.

I have left the code more or less the same except for the following changes:
1) move member initialisation to header file
2) cleanup headers
3) renderers now return which render capture class to allocate
4) corrected inheritance

More can be done to improve the code but I have left it as is for simplification for now. For the future it would be nice to allocate using `std::unique_ptr` instead of a raw pointer, this will require more changes in `CRenderManager` though. I would also like to use `std::vector` rather than a raw array for the pixel buffer.

This should be tested on both GL and DX platforms.

### CODE REVIEW

File: xbmc/cores/VideoPlayer/VideoRenderers/RenderCaptureDX.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#include "RenderCapture.h"
+#include "guilib/D3DResource.h"
+
+#include <wrl/client.h>
+
+class CRenderCaptureDX : public CRenderCapture, public ID3DResource
+{
+public:
+  CRenderCaptureDX();
+  ~CRenderCaptureDX() overide;
```
Comment (thexai) [inline xbmc/cores/VideoPlayer/VideoRenderers/RenderCaptureDX.h:20]:
```suggestion
  ~CRenderCaptureDX() override;
```
typo

File: xbmc/cores/VideoPlayer/VideoRenderers/RenderCaptureDX.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  void BeginRender() override;
+  void EndRender() override;
+  void ReadOut() overide;
```
Comment (thexai) [inline xbmc/cores/VideoPlayer/VideoRenderers/RenderCaptureDX.h:24]:
```suggestion
  void ReadOut() override;
```
typo

**Review by thexai — COMMENTED** [review]
On:
https://github.com/xbmc/xbmc/blob/2dbcd25bbf1051c26264d9a32087c77ed98719c0/xbmc/cores/VideoPlayer/VideoRenderers/WinRenderer.cpp#L228

capture->GetTarget()

`capture` object is from base class `CRenderCapture` and has not method `GetTarget()` maybe change to class `CRenderCaptureDX` here ?

These are compilation errors on Windows platforms, not tested yet.

```
06:20:28 C:\Users\Kodi\Kodi\jenkins\builds\workspace\WIN-32\xbmc\cores\VideoPlayer\VideoRenderers\WinRenderer.cpp(229): error C2039: 'GetTarget': is not a member of 'CRenderCapture' [C:\Users\Kodi\Kodi\jenkins\builds\workspace\WIN-32\kodi-build.x86\libkodi.vcxproj]
06:20:28   c:\users\kodi\kodi\jenkins\builds\workspace\win-32\xbmc\cores\videoplayer\videorenderers\WinRenderer.h(15): note: see declaration of 'CRenderCapture'RPWinOutputShader.cpp
```

**Review by thexai — APPROVED** [review]
Runtime tested on Windows x64. Tested bookmarks thumbnails and no issues found.

**Review by garbear — APPROVED** [review]
From a mechanical standpoint, it looks like the split was done correctly.

**Comment by lrusak** [comment]
I added a fixup that fixes a crash if no rendercapture is available (some renderers don't implement it).

**Comment by lrusak** [comment]
jenkins build this please

**Pull request merged**